pkg/controller/machinehealthcheck/machinehealthcheck_controller.go

```diff
@@ -264,6 +260,121 @@ func (r *ReconcileMachineHealthCheck) Reconcile(ctx context.Context, request rec
+		return errList
+	}
+
+	generateTemplateInput := &external.GenerateTemplateInput{
+		Template:    from,
+		TemplateRef: m.Spec.RemediationTemplate,
+		Namespace:   t.Machine.Namespace,
+		ClusterName: t.Machine.ClusterName,
+		OwnerRef:    cloneOwnerRef,
+	}
+	to, err := external.GenerateTemplate(generateTemplateInput)
+	if err != nil {
+		errList = append(errList, errors.Wrapf(err, "failed to create template for remediation request %v %q for machine %q in namespace %q within cluster %q", m.Spec.RemediationTemplate.GroupVersionKind(), m.Spec.RemediationTemplate.Name, t.Machine.Name, t.Machine.Namespace, m.Spec.ClusterName))
+		return errList
+	}
+
+	// Set the Remediation Request to match the Machine name, the name is used to
+	// guarantee uniqueness between runs. A Machine should only ever have a single
+	// remediation object of a specific GVK created.
+	//
+	// NOTE: This doesn't guarantee uniqueness across different MHC objects watching
+	// the same Machine, users are in charge of setting health checks and remediation properly.
+	to.SetName(t.Machine.Name)
+
+	klog.V(3).Infof("Target has failed health check, creating an external remediation request", "remediation request name", to.GetName(), "target", t.string(), "reason", condition.Reason, "message", condition.Message)
+	// Create the external clone.
+	if err := r.client.Create(ctx, to); err != nil {
+		conditions.MarkFalse(m, mapiv1.ExternalRemediationRequestAvailable, mapiv1.ExternalRemediationRequestCreationFailed, mapiv1.ConditionSeverityError, err.Error())
+		errList = append(errList, errors.Wrapf(err, "error creating remediation request for machine %q in namespace %q within cluster %q", t.Machine.Name, t.Machine.Namespace, t.Machine.ClusterName))
+		return errList
+	}
+	return errList
+}
+
+// getExternalRemediationRequest gets reference to External Remediation Request, unstructured object.
+func (r *ReconcileMachineHealthCheck) getExternalRemediationRequest(ctx context.Context, m *mapiv1.MachineHealthCheck, machineName string) (*unstructured.Unstructured, error) {
+	remediationRef := &corev1.ObjectReference{
+		APIVersion: m.Spec.RemediationTemplate.APIVersion,
+		Kind:       strings.TrimSuffix(m.Spec.RemediationTemplate.Kind, external.TemplateSuffix),
+		Name:       machineName,
+	}
+	remediationReq, err := external.Get(ctx, r.client, remediationRef, m.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve external remediation request object: %v", err)
+	}
+	return remediationReq, nil
+}
+
+// externalRemediationRequestExists checks if the External Remediation Request is created
+// for the machine.
+func (r *ReconcileMachineHealthCheck) externalRemediationRequestExists(ctx context.Context, m *mapiv1.MachineHealthCheck, machineName string) bool {
+	remediationReq, err := r.getExternalRemediationRequest(ctx, m, machineName)
+	if err != nil {
+		return false
```

Comment:
That's risky.
If there was an error, it doesn't mean the object doesn't exist.

Reply:
Good point.
Let's discuss the best approach

Reply:
I'd say the best approach is to take the code pieces from upstream, they already handle this I guess

Reply:
Well this is the upstream code.
let's discuss more over a meeting